fastlane_bot/events/pools/solidly_v2.py

```diff
@@ -88,11 +88,16 @@ def update_from_contract(
         """
         reserve_balance = contract.caller.getReserves()
 
-        try:
+        if hasattr(contract.caller, "factory"):
+            # in solidly, the factory is available via function `factory`
             factory_address = contract.caller.factory()
-        except Exception:
-            # Velocimeter does not expose factory function - call voter to get an address that is the same for all Velcoimeter pools
+        elif hasattr(contract.caller, "voter"):
+            # in velocimeter, the factory is available via function `voter`
             factory_address = contract.caller.voter()
+        else:
```

Comment:
shouldn't we check for cleopatra here rather than assuming it is and possibly continue with a completely wrong value?

Reply:
Correct.
Problem is, there is no way to tell whether or not the given contract is a Cleopatra pool.
BTW, the same problem exists prior to this PR (attempting to call contract function 'voter' as a fallback).
And due to the issue presented in a Stratum pool contract, where the address of the corresponding factory contract is not available in storage at all (neither via a contract function, nor by directly reading the contract storage), it seems that an overall different solution is required in the long run.

Reply:
Updated the `else` part such that if the `factory_address` is not the address of SOME contract (i.e., no bytecode deployed at that address), it is set to the zero-address.
Of course, there is not a huge gain in this, other than making potential problems easier to investigate.
First off, because we might end up with `factory_address` pointing to some non-factory contract, whose address happens to be stored in slot 11 of the pool contract.
Second, because a zero address will nevertheless throw an exception as soon as we attempt to interact with a factory contract "deployed" (i.e., obviously not deployed) at that address.
So again - we should probably conceive a different solution altogether for this entire feature of obtaining factory addresses (e.g., store them hard-coded somewhere).